R/weight.R

```diff
@@ -62,12 +123,105 @@ setMethod("weightVariables", "VariableCatalog", function (x) {
     ## Return the names/aliases
     if (length(vars)) {
         return(sort(vapply(index(vars), vget(namekey(x)), character(1),
-            USE.NAMES=FALSE)))
+            USE.NAMES = FALSE)))
     } else {
         return(c())
     }
 })
 
+#' @rdname weightVariables
+#' @export
+setMethod("weightVariables<-", "CrunchDataset", function (x, value) {
+    modifyWeightVariables(x, value, type = "append")
+    return(x)
+})
+
+
+#' modifyWeightVariables
+#'
+#' Change which variables can be set as a dataset's weight.
+#'
+#' @param x a CrunchDataset
+#' @param vars Variables to add or remove  this can be a numeric Crunch variable,
+#' list of numeric Crunch variables or a character vector with the aliases of numeric Crunch variables.
+#' @param type a character string determining how the weightVariables
+#' will be modified:
+#' - `"append"` : add `vars` to the current weight variables
+#' - `"remove"` : remove `vars` from the current list of weight variables
+#' - `"replace"`: replace the current weight variables with `vars`
+#' @return a CrunchDataset
+#' @export
+modifyWeightVariables <- function (x, vars, type = "append") {
+    varcat <- allVariables(x)
+    new <- old <- crGET(shojiURL(varcat, "orders", "weights"))
+    if (is.null(vars)) {
+        new$graph <- NULL
+        type <- "remove"
+    } else {
+        if (is.variable(vars) || (length(vars) == 1) & !is.character(vars)) {
+            vars <- list(vars)
+        }
+        if (is.character(vars)) {
+            ds <- loadDataset(datasetReference(x))
+            vars <- lapply(vars, function (v) {
+                if (v %in% names(varcat)) {
+                    ds[[v]]
+                } else {
+                    v
+                }
+            })
+        }
+        all_var <- vapply(vars, is.Numeric, logical(1))
+        if (!all(all_var)) {
+            var_names <- vapply(vars, function (v) {
+                if (is.variable(v)) {
+                    return(name(v))
+                } else {
+                    return(as.character(v))
+                }
+            }, character(1))
+            err_text <- " is not a numeric Crunch variable."
+            if (sum(!all_var) > 1) {
+                err_text <- " are not numeric Crunch variables."
+            }
+            halt(serialPaste(var_names[!all_var]), err_text)
+        }
+        if (type == "append") {
+            new$graph <- unique(c(old$graph, lapply(vars, self)))
+        } else if (type == "remove") {
+            new$graph <- setdiff(old$graph, lapply(vars, self))
+        } else if (type == "replace") {
+            new$graph <- lapply(vars, self)
+        }
+    }
+    if (!identical(old, new)) {
+        crPUT(shojiURL(varcat, "orders", "weights"),
+            body = toJSON(new))
+    }
+    return(x)
+}
+
+#' @rdname weightVariables
+#' @export
+is.weightVariable <- function (x) {
+    if (is.variable(x)) {
+        ds <- loadDataset(datasetReference(self(x)))
```

Comment:
Third time I've pointed out an unnecessary `self` inside of `datasetReference` calls. Please find and replace all. There are more than just this one left.